### ISSUE TITLE
Parsing issue on `BEGIN ATOMIC` functions containing `UNION ALL`

### ISSUE THREAD

**Issue by danny-andrews** [sqlfluff/sqlfluff#7847]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Follow-up to #7771. Most `BEGIN ATOMIC` functions now parse fine (thanks!) but there's one edge-case I found.

When attempting to lint the following function:

```sql
create or replace function union_things()
returns table (num integer)
language sql
stable
begin atomic
  select 1

  union all

  select 2;
end;
```

I get an error.

### Expected Behaviour

It parses without issue (this function is valid SQL).

### Observed Behaviour

I get `Position 7: Found unparsable section: 'atomic/n  select 1/n/n  union all/n/n  select...'sqlfluff(PRS)`.

### How to reproduce

Put the above function definition in a file and attempt to lint it.

### Dialect

postgres

### Version

Python: `3.14.4`
sqlfluff: `4.2.0`

### Configuration

pyproject.toml:

```toml
[project]
name = "repowr-db"
version = "0.0.1"
requires-python = "==3.14.4"

[dependency-groups]
dev = [
  "sqlfluff>=4.1.0",
]

[project.scripts]
lint = "scripts:lint"
lint-fix = "scripts:lint_fix"

[tool.setuptools.packages.find]
include = ["scripts"]

[tool.uv]
package = true

[tool.sqlfluff.core]
dialect = "postgres"
max_line_length = 80
rules = ["all"]
exclude_rules = [
  # Don't require order by for all limit queries.
  "ambiguous.order_by_limit",
  # Don't require newline after CTE definition.
  "layout.cte_newline",

  # Don't require qualification on all columns.
  "references.qualification",
  "references.consistent",
  # Allow keyword identifiers if they don't cause errors.
  "references.keywords",
  # Allow subqueries in from and joins.
  "structure.subquery",
  # Allow using clause.
  "structure.using",
  # Allow aggregate columns to come in any order.
  "structure.column_order",
  "references.special_chars",
]

[tool.sqlfluff.indentation]
tab_space_size = 2

[tool.sqlfluff.rules.convention]
not_equal.preferred_not_equal_style = "c_style"
terminator.require_final_semicolon = true
casting_style.preferred_type_casting_style = "shorthand"
blocked_words.blocked_words = "outer"

[tool.sqlfluff.rules.capitalisation]
keywords.capitalisation_policy = "lower"
identifiers.extended_capitalisation_policy = "lower"
functions.extended_capitalisation_policy = "lower"
literals.capitalisation_policy = "lower"
types.extended_capitalisation_policy = "lower"

[tool.sqlfluff.rules.structure]
join_condition_order.preferred_first_table_in_join_clause = 'later'

[tool.sqlfluff.rules.layout]
newlines.maximum_empty_lines_between_statements = 1
```

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by danny-andrews** [comment]
A workaround is to use the `return`-style:

```sql
-- This parses and is equivalent
create or replace function union_things()
returns table (num integer)
language sql
stable
return
  select 1

  union all

  select 2;
```
